Productivity_Analysis_Data.R

```diff
@@ -1,42 +1,69 @@
+# Load libraries
+if (!'pacman' %in% installed.packages()[,'Package']) install.packages('pacman', repos='http://cran.r-project.org')
+pacman::p_load(dplyr,plyr,zoo)
 
+# Define "not in" operator
+'%nin%' <- Negate('%in%')
 
-## ------------------------------------------------------------------------
-## Clean up the data and extract relevant variables
-library(plyr)
-library(dplyr)
-library(zoo)
+# Define country names and years
+year_names <- c(2006:2015)
+country_names  <- c('Albania', 'Austria', 'Belarus', 'Belgium', 'Bosnia and Herzegovina', 'Bulgaria', 'Croatia', 'Cyprus', 'Czech Republic', 'Denmark', 'Estonia', 'Finland', 'France', 'Germany', 'Greece', 'Hungary', 'Iceland', 'Ireland', 'Italy', 'Kosovo', 'Latvia', 'Liechtenstein', 'Lithuania', 'Luxembourg', 'Macedonia, FYR', 'Malta', 'Monaco', 'Montenegro', 'Netherlands', 'Norway', 'Poland', 'Portugal', 'Moldova', 'Romania', 'Russian Federation', 'Serbia', 'Slovakia', 'Slovenia', 'Spain', 'Sweden', 'Switzerland', 'Turkey', 'Ukraine', 'United Kingdom')
 
-non_na_ind <- list(); Firm_Size_list <- list(); Index_list <- list(); Productivity_list <- list(); Profitability_list <- list(); Cost_Structure_list <- list()
+# Prepare lists to collect country data
+non_na_ind <- list()
+Firm_Size_list <- list()
+Index_list <- list()
+Productivity_list <- list() 
+Profitability_list <- list()
+Cost_Structure_list <- list()
 
-for(i in c(1:44)){
+# Loop over countries, read files and collect desired data into lists
+for(i in c(1:length(country_names))){
   print(i)
   file_name <- paste("panels_J!&", paste(country_names[i], collapse=""), ".Rda", sep="")
   load(file_name)
   
-  non_na_ind[[i]] <- which(apply(is.na(Cleaned_dat_Firm_Size[,c(4,5,7)]), 1, sum) == 0) # non-missing for emp, toas, and va 
+  # get indices
+  idx_required_present_Firm_size_variables <- which(colnames(Cleaned_dat_Firm_Size) %in% c("EMPL","TOAS", "VA"))
+  idx_retain_INDEX_variables <- which(colnames(Cleaned_dat_INDEX) %in% c("NUTS_1", "NUTS_2", "NUTS_3", "NACE_PRIM_CODE", "COMPCAT", "Firm_Age"))
+  idx_retain_Productivity_variables <- which(colnames(Cleaned_dat_Productivity) %nin% c("IDNR", "Year"))
+  idx_retain_Profitability_variables <- which(colnames(Cleaned_dat_Profitability) %nin% c("IDNR", "Year"))
+  idx_retain_Cost_Structure_variables <- which(colnames(Cleaned_dat_Cost_Structure) %in% c("WS", "WS_AD"))
   
-  Firm_Size_list[[i]] <- Cleaned_dat_Firm_Size[non_na_ind[[i]],] 
   
-  Index_list[[i]] <- Cleaned_dat_INDEX[non_na_ind[[i]], c(5,6,8,10,13)]# nut_3, nace code, firm_size, firm_age, and name
+  non_na_ind[[i]] <- which(apply(is.na(Cleaned_dat_Firm_Size[, idx_required_present_Firm_size_variables]), 1, sum) == 0) # non-missing for emp, toas, and va 
```

Comment:
The following line should be more than 50% faster (also simpler):

non_na_ind[[i]] <- which(complete.cases(Cleaned_dat_Firm_Size[, idx_required_present_Firm_size_variables]))